gadget/gadget.go

```diff
@@ -400,15 +416,76 @@ func validateVolume(name string, vol *Volume) error {
 			knownFsLabels[s.Label] = true
 		}
 
+		switch s.Role {
+		case SystemSeed:
+			if state.SystemSeed != nil {
+				return fmt.Errorf("cannot have more than one system-seed role")
+			}
+			state.SystemSeed = &vol.Structure[idx]
+		case SystemData:
+			if state.SystemData != nil {
+				return fmt.Errorf("cannot have more than one system-seed role")
+			}
+			state.SystemData = &vol.Structure[idx]
+		}
+
 		previousEnd = end
 	}
 
+	if err := ensureVolumeConsistency(state, constraints); err != nil {
+		return err
+	}
+
 	// sort by starting offset
 	sort.Sort(byStartOffset(structures))
 
 	return validateCrossVolumeStructure(structures, knownStructures)
 }
 
+func ensureVolumeConsistency(state *validationState, constraints *ModelConstraints) error {
+	if constraints == nil {
+		if state.SystemData != nil {
+			// gadget must be auto-consistent if constraints are not specified
+			if state.SystemSeed != nil {
+				if state.SystemData.Label != UbuntuSystemDataLabel {
+					return fmt.Errorf("system-data structure must have label %q, not %q",
+						UbuntuSystemDataLabel, state.SystemData.Label)
+				}
+			} else {
+				if state.SystemData.Label != "" && state.SystemData.Label != ImplicitSystemDataLabel {
+					return fmt.Errorf("system-data structure must have an implicit label or %q, not %q",
+						ImplicitSystemDataLabel, state.SystemData.Label)
+				}
+			}
+		}
+		return nil
+	}
+
+	if state.SystemSeed != nil {
+		// error if we don't have the SystemSeed constraint but we have a system-seed structure
+		if !constraints.SystemSeed {
+			return fmt.Errorf("model does not support the system-seed role")
+		}
+		// with SystemSeed, system-data label must be ubuntu-data
+		if state.SystemData != nil && state.SystemData.Label != UbuntuSystemDataLabel {
```

Comment:
I don't know about previous agreements, do we expect the label to be explicitly provided?

Reply:
W need to chat briefly about this with @mvo5, but I do think that given other choices we made supporting  implicit labels for these would still make sense.

Reply:
Yes, if we're in system-seed mode, the ubuntu-data label should be explicit. Is this still our current understanding, @mvo5 ?

Reply:
@pedronis Ok, waiting to check this. If implicit system-data labels are to be used I'll have to adjust the code a little bit since the current validation won't allow this.

Reply:
Changed validation rules so that system-data must not have a label if system-seed is in use.